### PR TITLE
Fix constraint issues in connector generation

### LINKED ISSUE #1053: Missing type in Generated `type.bal` when anyOf has constraint scenarios
**Description:**
```openapi

components:
  schemas:
    User:
      properties:
        first_name:
          type: string
          description: First name of the Uber user.
        last_name:
          type: string
          description: Last name of the Uber user.
        tax_rates:
          anyOf:
            - items:
                maxLength: 5000
                type: string
              type: array
            - enum:
                - ''
              type: string
```
generated code 
```ballerina 
public type User record {
    # First name of the Uber user.
    string first_name?;
    # Last name of the Uber user.
    string last_name?;
    TaxratesItemsString[]|string tax_rates?;
};
```
expected code
```ballerina
import ballerina/constraint;
@constraint:String{maxLength: 5000}
public type TaxratesItemsString string;
public type User record {
    # First name of the Uber user.
    string first_name?;
    # Last name of the Uber user.
    string last_name?;
    TaxratesItemsString[]|string tax_rates?;
};
```
**Steps to reproduce:**

**Affected Versions:**

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### LINKED ISSUE #1054: Generate constraint gives compilation issue when the maxValue, minValue have decimal or float value
**Description:**
Wrong code generations 
```ballerina
# The properties of the messaging endpoints used by this IoT hub.
public type MessagingEndpointProperties record {
    # The lock duration. See: https://docs.microsoft.com/azure/iot-hub/iot-hub-devguide-file-upload.
    string lockDurationAsIso8601?;
    # The number of times the IoT hub attempts to deliver a message. See: https://docs.microsoft.com/azure/iot-hub/iot-hub-devguide-file-upload.
    @constraint:Int {minValue: 1, maxValue: 100.0}
    int maxDeliveryCount?;
    # The period of time for which a message is available to consume before it is expired by the IoT hub. See: https://docs.microsoft.com/azure/iot-hub/iot-hub-devguide-file-upload.
    string ttlAsIso8601?;
};
```

**Steps to reproduce:**

**Affected Versions:**

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### LINKED ISSUE #1056: Schema generation has compilation issue when the union type has constraint support
**Description:**
Currently, constraint support is not available with the union type , therefore generated schema has a compilation issue when the below scenarios
```ballerina 
public type IssuingTransactionFlightDataLeg record {
    # The three-letter IATA airport code of the flight's destination.
    @constraint:String {maxLength: 5000}
    string? arrival_airport_code?;
    # The airline carrier code.
    @constraint:String {maxLength: 5000}
    string? carrier?;
    # The three-letter IATA airport code that the flight departed from.
    @constraint:String {maxLength: 5000}
    string? departure_airport_code?;
    # The flight number.
    @constraint:String {maxLength: 5000}
    string? flight_number?;
    # The flight's service class.
    @constraint:String {maxLength: 5000}
    string? service_class?;
    # Whether a stopover is allowed on this flight.
    boolean? stopover_allowed?;
};
```
compilation issue
```
ERROR [types.bal:(223:5,223:41)] invalid `@constraint:String` annotation on `string?` type
ERROR [types.bal:(230:5,230:41)] invalid `@constraint:String` annotation on `string?` type
ERROR [types.bal:(233:5,233:41)] invalid `@constraint:String` annotation on `string?` type
ERROR [types.bal:(236:5,236:41)] invalid `@constraint:String` annotation on `string?` type
ERROR [types.bal:(239:5,239:41)] invalid `@constraint:String` annotation on `string?` type
ERROR [types.bal:(242:5,242:41)] invalid `@constraint:String` annotation on `string?` type
.....
````

**Steps to reproduce:**

**Affected Versions:**

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by lnash94**
Fix #1054 :  Generate constraint gives compilation issue when the maxValue, minValue have decimal or float value
Fix #1056 :  Schema generation has compilation issue when the union type has constraint support

### CODE REVIEW

File: openapi-cli/src/main/java/io/ballerina/openapi/cmd/OpenApiCmd.java [hunk 1]
```diff
@@ -267,6 +267,9 @@ private void openApiToBallerina(String fileName, Filter filter) throws IOExcepti
         }
         getTargetOutputPath();
         Path resourcePath = Paths.get(openApiFile.getCanonicalPath());
+        if (nullable) {
+            outStream.println("WARNING: `@constraint` support will not be available with `--nullable` option");
```
Comment (NipunaRanasinghe) [inline openapi-cli/src/main/java/io/ballerina/openapi/cmd/OpenApiCmd.java:271]:
```suggestion
            outStream.println("WARNING: All the constraints in OpenAPI contract will be ignored when generating the Ballerina client/service with `--nullable` option");
```

File: openapi-cli/src/main/java/io/ballerina/openapi/generators/schema/BallerinaTypesGenerator.java [hunk 2]
```diff
@@ -207,11 +209,33 @@ public TypeDefinitionNode getTypeDefinitionNode(Schema schema, String typeName,
...
+        } else if (typeGenerator instanceof UnionTypeGenerator &&
+                !((UnionTypeGenerator) typeGenerator).getTypeDefinitionNodeList().isEmpty()) {
+            removeDuplicateNode(((UnionTypeGenerator) typeGenerator).getTypeDefinitionNodeList());
         }
         return typeDefinitionNode;
     }
+
+    /**
+     * Remove duplicate of the TypeDefinitionNode.
+     *
```
Comment (NipunaRanasinghe) [inline openapi-cli/src/main/java/io/ballerina/openapi/generators/schema/BallerinaTypesGenerator.java:225]: Additional blank line

File: openapi-cli/src/main/java/io/ballerina/openapi/generators/schema/TypeGeneratorUtils.java [hunk 3]
```diff
@@ -179,6 +176,19 @@ public static void updateRecordFieldList(List<String> required,
         //Generate constraint annotation.
         AnnotationNode constraintNode = generateConstraintNode(fieldSchema);
         MetadataNode metadataNode;
+        boolean isConstraintSupport =
```
Comment (NipunaRanasinghe) [inline openapi-cli/src/main/java/io/ballerina/openapi/generators/schema/TypeGeneratorUtils.java:179]: Aren't we checking if this schema belongs to a union type? If so, shall we rename as follows?
```suggestion
        boolean isUnionType =
```

File: openapi-cli/src/main/java/io/ballerina/openapi/generators/schema/TypeGeneratorUtils.java [hunk 4]
```diff
@@ -262,6 +272,7 @@ public static AnnotationNode generateConstraintNode(Schema<?> fieldSchema) {
             return generateArrayConstraint(arraySchema);
         }
         // Ignore Object, Map and Composed schemas.
+
```
Comment (NipunaRanasinghe) [inline openapi-cli/src/main/java/io/ballerina/openapi/generators/schema/TypeGeneratorUtils.java:275]: Additional newline

File: openapi-cli/src/main/java/io/ballerina/openapi/generators/schema/TypeGeneratorUtils.java [hunk 5]
```diff
@@ -179,6 +176,19 @@ public static void updateRecordFieldList(List<String> required,
...
+                constraintNode != null && fieldSchema.getNullable() != null && fieldSchema.getNullable() ||
+                (fieldSchema instanceof ComposedSchema && (((ComposedSchema) fieldSchema).getOneOf() != null ||
+                                ((ComposedSchema) fieldSchema).getAnyOf() != null));
+        boolean nullable = GeneratorMetaData.getInstance().isNullable();
+        if (nullable) {
+            constraintNode = null;
+        } else if (isConstraintSupport) {
+            PrintStream outStream = System.out;
+            outStream.println(String.format("WARNING: `@constraint` support will not be available with " +
+                    "union type in given `%s` field", fieldName.toString().trim()));
```
Comment (NipunaRanasinghe) [inline openapi-cli/src/main/java/io/ballerina/openapi/generators/schema/TypeGeneratorUtils.java:189]:
```suggestion
            outStream.println(String.format("WARNING: constraints in OpenAPI contract will be ignored for the field  `%s`, as constraints are not supported on Ballerina union types", fieldName.toString().trim()));
```

File: openapi-cli/src/test/java/io/ballerina/openapi/generators/schema/ConstraintTests.java [hunk 6]
```diff
@@ -103,6 +103,15 @@ public void testRecordFiledConstraintWithZeroValue() throws IOException, Balleri
...
     }
+    /**
+     * example warnings for union type has constraint
+     *
+     * WARNING: `@constraint` support will not be available with union type in given `service_class` field
+     * WARNING: `@constraint` support will not be available with union type in given `tax_rates` field
+     * WARNING: `@constraint` support will not be available with union type in given `tax_rates_anyOf` field
+     * WARNING: `@constraint` support will not be available with union type in given `tax_rates_oneOF_array` field
+     * WARNING: `@constraint` support will not be available with union type in given `tax_rates_anyOf_array` field
+     */
```
Comment (NipunaRanasinghe) [inline openapi-cli/src/test/java/io/ballerina/openapi/generators/schema/ConstraintTests.java:114]: Let's add negative test cases to ensure that all the above unsupported scenarios are handled by the tool.

**Review by NipunaRanasinghe — CHANGES_REQUESTED** [review]

File: openapi-cli/src/test/java/io/ballerina/openapi/generators/schema/ConstraintTests.java (hunk 6 above)
Comment (lnash94) [reply openapi-cli/src/test/java/io/ballerina/openapi/generators/schema/ConstraintTests.java]: I have added with the same test, will separate the test with warnings  :)

File: openapi-cli/src/test/java/io/ballerina/openapi/generators/schema/ConstraintTests.java [hunk 7]
```diff
@@ -26,21 +26,31 @@
...
  */
 public class ConstraintTests {
     private static final Path RES_DIR = Paths.get("src/test/resources/generators/schema").toAbsolutePath();
     CodeGenerator codeGenerator = new CodeGenerator();
+    private final ByteArrayOutputStream outContent = new ByteArrayOutputStream();
+    private final PrintStream originalOut = System.out;
+    @BeforeTest
+    public void setUpStreams() {
+        System.setOut(new PrintStream(outContent));
+    }
```
Comment (NipunaRanasinghe) [inline openapi-cli/src/test/java/io/ballerina/openapi/generators/schema/ConstraintTests.java:53]: Missing new lines before and after the method

**Review by NipunaRanasinghe — APPROVED** [review]

**Pull request merged**